# RNA-seq Differential Expression Analysis

## Background 

This project aimed to identify genes that were differentially expressed in the amygdala and hippocampus between two closely related inbred mouse strains, and between social strata of an idiopathic mouse model for autism spectrum disorder. RNA was extracted from tissue punches of the respective regions and libraries were generated using the Illumina TrueSeq mRNA LP kit and sequenced on a NovaSeq 6000 S1 full flowcell. 

#### Overall workflow
- QC-filtering/adapter trimming (cutadapt, FASTQC, bash)
- Alignment (STAR - GRCm38.96, bash)
- Generate counts for known transcripts (FeatureCount, bash)
- Principal component analysis (PCAtools, R)
- Differential expression analysis (DESeq2, R)
- Gene set enrichment analysis (fGSEA, R)

## Description of contents
- C58_C57_expanded_models_results.Rmd: Performs PCA on the expression matrices generated by FeatureCount, differential expression analysis using DESeq2, BioMart to annotate results (gene ID conversion, biotype annotation), visualization and summary tables (volcano plots, boxplots, results tables, results overview report, etc...).
- fGSEA_rnaseq.Rmd: Perform gene set enrichment analyses using the test statistic from the DESeq2 analyses with fgsea and four pathways provided by msigdbr (Gene ontology terms (biological pathways, cellular components, molecular functions), and the Hallmark gene set).  
- DESeq2_heatmaps.Rmd: visualizing the top 200 differentially expressed genes and the hierarchical clustering patterns of the mouse samples. 

## Summary results

A heatmap depicting the 24 samples included in the differential expression analyses of the hippocampus. Pictured are the top 200 differentially expressed genes between strains. 

![](https://github.com/stephen-siecinski/social_behavior_in_mice/blob/main/rna-seq/c58-b6_heatmap_hippocampus_protein-coding_20210226.png?raw=true)
